### PR TITLE
More integrated circuit fixes and code improvements

### PR DESCRIPTION
**Pull request by ACCount12**
Lots of tweaks, fixes, code improvements and UI improvements.

Also includes one new component type: list constructor/deconstructor. Useful if you want to turn a list into a bunch of values without using a million of "at" circuits.

* Freshly printed IC assemblies now start with their panels open.
* Clicking open assemblies with multitool now opens the UI correctly.
* Fixes runtimes caused by lack of sanity checks on list components.
* Fixes "pick" component only working on strings.
* Fixes text/char pins trimming user input, making setting " " as separator unnecessary hard (includes #32961).
* Fixes multiple thrower component bugs and runtime errors.
* Assembly UI has a new, much more convenient set of controls for rearranging components:
default
* Component UI now only displays "Remove" and "Return to Assembly" options if it's actually located in an assembly.
* Component UI is now much more likely to be valid HTML.
* LEDs now support RGB and no longer clutter the list of components.
* Assemblies now use `assembly_components` list instead of looping for components in contents.
* Code responsible for removing or rearranging assembly components is moved to assemblies.
* All material costs are now calculated using the same logic.
* Two massive integrated circuit sprite files are split into a bunch of smaller ones.
* Split and join components now have their own sprites.
* New pin type: index pin. Only accepts integers between 1 and IC_MAX_LIST_LENGTH, useful if you don't want to have too many sanity checks in your list components.
* New helpers added for connecting and disconnecting pins, adding and removing assembly components,  tools being used on pins and more.

:cl: ACCount
add: New integrated circuit components: list constructors/deconstructors. Useful for building lists and taking them apart.
fix: Fixed multiple bugs in integrated circuits UIs, improved overall usability.
/:cl:

### CODE REVIEW

**Comment by ACCount12** [comment]
I'm pretty much done with this one. Unless I stumble upon a massive unfixed bug any time soon, this is the final contents of this PR.

File: code/modules/integrated_electronics/core/assemblies.dm [hunk 1]
```diff
@@ -40,11 +41,12 @@
 
 /obj/item/device/electronic_assembly/proc/handle_idle_power()
 	// First we generate power.
-	for(var/obj/item/integrated_circuit/passive/power/P in contents)
+	for(var/obj/item/integrated_circuit/passive/power/P in components)
```
Comment (AnturK) [inline code/modules/integrated_electronics/core/assemblies.dm]: Could you name this circuit_components or assembly_components ? So it's more obvious these are not generic ones.

File: code/modules/integrated_electronics/subtypes/manipulation.dm [hunk 2]
```diff
@@ -362,26 +362,39 @@
...
-	if(A.anchored || (A.w_class > WEIGHT_CLASS_NORMAL))
+
+	var/atom/movable/acting_object = get_object()
+	var/turf/T = get_turf(acting_object)
+	if(!T)
 		return
-	var/turf/T = get_turf(assembly)
-	if(!(A.Adjacent(T) || (A in assembly.GetAllContents())))
+
+	if(!A.Adjacent(T) && !(A in acting_object.GetAllContents()))
```
Comment (AnturK) [inline code/modules/integrated_electronics/subtypes/manipulation.dm]: Once again, why compare against the turf if you know the holder/acting object.

**Review by AnturK — CHANGES_REQUESTED** [review]

**Comment by ACCount12** [comment]
I changed what was requested.

**Pull request merged**